core/services/vrf/listener_v2.go

```diff
@@ -306,11 +348,234 @@ func (a fulfilledReqV2) Compare(b heaps.Item) int {
+			}
+
+			proofs = append(proofs, batch_vrf_coordinator_v2.VRFTypesProof(pf))
+			commitments = append(commitments, batch_vrf_coordinator_v2.VRFTypesRequestCommitment(rc))
+			runs = append(runs, run)
+			reqIDs = append(reqIDs, vrfRequest.RequestId)
+			lbs = append(lbs, req.lb)
+			maxLinks = append(maxLinks, maxLink)
+			totalGasLimit += gaslimit
+
+			startBalanceNoReserveLink.Sub(startBalanceNoReserveLink, maxLink)
+		}
+
+		// If we didn't enqueue anything into any of the slices
+		// then we broke out of the above loop due to out of funds.
+		// Break out of this outer loop since there's nothing left to do.
+		if len(proofs) == 0 {
+			lggr.Info("Nothing left to process in batch")
+			break
+		}
+
+		// Enqueue a single batch tx for requests that we're able to fulfill based on whether
+		// they passed simulation or not.
+		payload, err := batchCoordinatorV2ABI.Pack("fulfillRandomWords", proofs, commitments)
+		if err != nil {
+			// should never happen
+			lggr.Errorw("Failed to pack batch fulfillRandomWords payload",
+				"err", err, "proofs", proofs, "commitments", commitments)
+			continue
+		}
+
+		// Bump the total gas limit by a bit so that we account for the overhead of the batch
+		// contract's calling.
+		totalGasLimitBumped := uint64(float64(totalGasLimit) * 1.15)
+		lggr.Infow("Enqueueing batch fulfillment",
+			"numRequestsInBatch", len(reqIDs),
+			"requestIDs", reqIDs,
+			"totalGasLimit", totalGasLimit,
+			"linkBalance", startBalanceNoReserveLink,
+			"totalGasLimitBumped", totalGasLimitBumped,
+		)
+		err = lsn.q.Transaction(func(tx pg.Queryer) error {
+			for j := range runs {
+				if err = lsn.pipelineRunner.InsertFinishedRun(&runs[j], true, pg.WithQueryer(tx)); err != nil {
+					return err
+				}
+			}
+
+			for _, lb := range lbs {
+				if err = lsn.logBroadcaster.MarkConsumed(lb, pg.WithQueryer(tx)); err != nil {
+					return err
+				}
+			}
+
+			maxLinkStr := bigmath.Accumulate(maxLinks).String()
+			reqIDHashes := []common.Hash{}
+			for _, reqID := range reqIDs {
+				reqIDHashes = append(reqIDHashes, common.BytesToHash(reqID.Bytes()))
+			}
+			_, err = lsn.txm.CreateEthTransaction(txmgr.NewTx{
+				FromAddress:      fromAddress,
+				ToAddress:        lsn.batchCoordinator.Address(),
+				EncodedPayload:   payload,
+				GasLimit:         totalGasLimitBumped,
+				MinConfirmations: null.Uint32From(uint32(lsn.cfg.MinRequiredOutgoingConfirmations())),
+				Strategy:         txmgr.NewSendEveryStrategy(),
+				Meta: &txmgr.EthTxMeta{
+					RequestIDs: reqIDHashes,
+					MaxLink:    &maxLinkStr,
+					SubID:      &subID,
+				},
+				// TODO: implement checker for batches.
+				// Checker: txmgr.TransmitCheckerSpec{
+				// 	CheckerType:           txmgr.TransmitCheckerTypeVRFV2,
+				// 	VRFCoordinatorAddress: lsn.coordinator.Address(),
+				// },,
+			}, pg.WithQueryer(tx))
+
+			return err
+		})
+		if err != nil {
+			lggr.Errorw("Error enqueuing batch fulfillments, requeuing requests", "err", err)
+			continue
```

Comment:
I'm not sure this in particular could cause an infinite loop, since at this point `i` is presumably incremented to some new value, and after `continue` we'd check if `i` is still a valid index and either execute the outer loop again or break from it if it's too big, unless I'm missing something.

Reply:
yeah it seems ok as is, just brittle. Let's document why this continue is safe at least

Reply:
I've changed the loop setup to be more like the current parallelized `processRequestsForSub` which processes in chunks, so this is no longer a factor.